src/Test/ASTNodeTestCase.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+/*
+ * This file is part of Twig.
+ *
+ * (c) Fabien Potencier
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Twig\Test;
+
+use PHPUnit\Framework\TestCase;
+use Twig\Compiler;
+use Twig\Environment;
+use Twig\Loader\ArrayLoader;
+use Twig\Node\Node;
+
+abstract class ASTNodeTestCase extends TestCase
+{
+    abstract public static function getTests();
+
+    /**
+     * @dataProvider getTests
+     */
+    public function testCompile($node, $source, $environment = null, $isPattern = false)
+    {
+        static::assertNodeCompilation($source, $node, $environment, $isPattern);
+    }
+
+    public static function assertNodeCompilation($source, Node $node, Environment $environment = null, $isPattern = false)
```

Comment:
Could you check if this method needs to be static, same for `getCompiler` and `getEnvironment`?

Reply:
keeping `getEnvironment` non-static would be more flexible (test cases could then use mocks in it if they need it). so I would keep all these non-static

Reply:
When doing so, you cannot use them in data providers anymore. I thought it would be more valueable to allow them to be used in data providers.
We use `getCompiler` in one test internally.

I don't have a strong opinion on this though. 

I've pushed a commit turning this non-static again.

Reply:
`\Twig\Tests\Node\DeprecatedTest` uses `Compiler::getVarName()` to build the expected php source code that is compiled from a `Node`.

I suppose we should rather hardcode `__internal_compile_0` into the expected string. What do you think?
I pushed a commit doing that on top of this branch in case you agree.